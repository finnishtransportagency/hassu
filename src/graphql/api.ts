import { Amplify } from "@aws-amplify/core";

import awsExports from "../aws-exports";
import { API, graphqlOperation } from "aws-amplify";
import { GraphQLAPIClass } from "@aws-amplify/api-graphql";
import * as model from "./apiModel";
import * as queries from "./queries";
import * as mutations from "./mutations";
import log from "loglevel";

import { GraphQLResult } from "@aws-amplify/api-graphql/src/types/index";

Amplify.configure(awsExports);

const yllapitoEndpoint = awsExports.aws_appsync_graphqlEndpoint.replace("/graphql", "/yllapito/graphql");
export const yllapitoAPI = new GraphQLAPIClass({ ...awsExports, aws_appsync_graphqlEndpoint: yllapitoEndpoint });
Amplify.register(yllapitoAPI);
yllapitoAPI.configure({ aws_appsync_graphqlEndpoint: yllapitoEndpoint });

const callYllapitoAPI = async (operation: any) => {
  try {
    return yllapitoAPI.graphql(operation);
  } catch (e) {
    log.error(e);
    window.location.pathname = "/yllapito/kirjaudu";
  }
};

export async function getSuunnitelmaById(suunnitelmaId: string) {
  const response = (await callYllapitoAPI(
    graphqlOperation(queries.getSuunnitelmaById, { suunnitelmaId })
  )) as GraphQLResult<model.GetSuunnitelmaByIdQuery>;
  return response.data?.getSuunnitelmaById as model.Suunnitelma;
}

export async function createSuunnitelma(suunnitelma: model.CreateSuunnitelmaInput) {
  return await callYllapitoAPI(graphqlOperation(mutations.createSuunnitelma, { suunnitelma }));
}

export async function updateSuunnitelma(suunnitelma: model.UpdateSuunnitelmaInput) {
  return await callYllapitoAPI(graphqlOperation(mutations.updateSuunnitelma, { suunnitelma }));
}

export async function getVelhoSuunnitelmasByName(
  suunnitelmaName: string,
  requireExactMatch?: boolean
): Promise<model.VelhoHakuTulos[]> {
  const response = (await callYllapitoAPI(
    graphqlOperation(queries.getVelhoSuunnitelmasByName, { suunnitelmaName, requireExactMatch })
  )) as GraphQLResult<model.GetVelhoSuunnitelmasByNameQuery>;
  return response.data?.getVelhoSuunnitelmasByName as model.VelhoHakuTulos[];
}

export async function listSuunnitelmat() {
  const response = (await API.graphql(
    graphqlOperation(queries.listSuunnitelmat)
  )) as GraphQLResult<model.ListSuunnitelmatQuery>;
  return response.data?.listSuunnitelmat as model.Suunnitelma[];
}

export async function getCurrentUser(): Promise<model.User | undefined> {
  try {
    const response = (await callYllapitoAPI(
      graphqlOperation(queries.getCurrentUser)
    )) as GraphQLResult<model.GetCurrentUserQuery>;
    return response.data?.getCurrentUser as model.User;
  } catch (e) {
    log.debug(e);
  }
}
