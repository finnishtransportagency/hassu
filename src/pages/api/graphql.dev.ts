import type { NextApiRequest, NextApiResponse } from "next";
import { HttpRequest } from "@aws-sdk/protocol-http";
import { HeaderBag } from "@aws-sdk/types";
import { sendSignedRequest } from "../../../backend/src/aws/awsRequest";

export default async function handler(req: NextApiRequest, res: NextApiResponse) {
  let body = req.body;
  const urlObject = new URL(process.env.APPSYNC_URL || "");
  const isYllapito = req.query.yllapito == "true";

  let headers = {
    host: urlObject.host,
  } as unknown as HeaderBag;
  if (isYllapito) {
    headers["x-hassudev-uid"] = getCookieOrDefault(req.cookies, "x-hassudev-uid", process.env["x-hassudev-uid"]) as string;
    headers["x-hassudev-roles"] = getCookieOrDefault(req.cookies, "x-hassudev-roles", process.env["x-hassudev-roles"]) as string;
  }
  const request = new HttpRequest({
    headers: headers,
    hostname: urlObject.hostname,
    method: "POST",
    path: urlObject.pathname,
    body: JSON.stringify(body),
  });
  const { body: responseBody, statusCode } = await sendSignedRequest(request, "appsync");

  await res.status(statusCode).json(responseBody);
}

function getCookieOrDefault(cookies: Partial<{ [p: string]: string }>, name: string, defaultValue: string | undefined) {
  let value = cookies?.[name];
  if (value) {
    return value;
  }
  return defaultValue;
}
