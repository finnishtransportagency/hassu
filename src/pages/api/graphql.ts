import type { NextApiRequest, NextApiResponse } from "next";
import { HttpRequest } from "@aws-sdk/protocol-http";
import { HeaderBag } from "@aws-sdk/types";
import { sendSignedRequest } from "../../../backend/src/aws/awsRequest";

export default async function handler(req: NextApiRequest, res: NextApiResponse) {
  let body = req.body;
  const urlObject = new URL(process.env.APPSYNC_URL || "");
  const isYllapito = req.query.yllapito == "true";

  let params: URLSearchParams | undefined;
  if (typeof window !== "undefined") {
    params = window.location?.search ? new URLSearchParams(window.location.search) : undefined;
  }
  let headers = {
    host: urlObject.host,
  } as unknown as HeaderBag;
  if (isYllapito) {
    headers["x-hassudev-uid"] = getParamOrDefault(params, "x-hassudev-uid", process.env["x-hassudev-uid"]) as string;
    headers["x-hassudev-roles"] = getParamOrDefault(
      params,
      "x-hassudev-roles",
      process.env["x-hassudev-roles"]
    ) as string;
  }
  const request = new HttpRequest({
    headers: headers,
    hostname: urlObject.hostname,
    method: "POST",
    path: urlObject.pathname,
    body: JSON.stringify(body),
  });
  const { body: responseBody, statusCode } = await sendSignedRequest(request, "appsync");

  res.status(statusCode).json(JSON.stringify(responseBody));
}

function storeValue(name: string, value: string) {
  if (typeof window !== "undefined") {
    localStorage.setItem(name, value);
  }
}

function getValue(name: string) {
  if (typeof window !== "undefined") {
    return localStorage.getItem(name);
  }
}

function getParamOrDefault(params: URLSearchParams | undefined, name: string, defaultValue: string | undefined) {
  if (params) {
    if (params.has(name)) {
      const value = params.get(name) || "";
      storeValue(name, value);
      return value;
    }
  }
  const valueFromStorage = getValue(name);
  if (!!valueFromStorage) {
    return valueFromStorage;
  }
  return defaultValue;
}
