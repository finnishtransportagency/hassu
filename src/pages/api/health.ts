import { NextApiRequest, NextApiResponse } from "next";

export default async function handler(_req: NextApiRequest, res: NextApiResponse) {
  res.status(200).json({ status: "ok" });
}
