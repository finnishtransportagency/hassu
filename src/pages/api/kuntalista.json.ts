import type { NextApiRequest, NextApiResponse } from "next";
import { S3Cache } from "../../../backend/src/cache/s3Cache";
import log from "loglevel";
import { setupLambdaMonitoring, wrapXrayAsync } from "../../../backend/src/aws/monitoring";

const KUNTALISTA_TTL_SECONDS = 24 * 3 * 60 * 60;

async function fetchKuntaLista() {
  const response = await fetch("http://rajapinnat.ymparisto.fi/api/Hakemistorajapinta/1.0/odata/Kunta");

  const data: any = await response.json();
  const list = data?.value?.map((kunta: any) => {
    return { label: kunta.Nimi, value: kunta.Nimi.toUpperCase() };
  });
  list.unshift({ label: "", value: "" });
  return list;
}

export default async function handler(_req: NextApiRequest, res: NextApiResponse) {
  setupLambdaMonitoring();
  return await wrapXrayAsync("handler", async () => {
    const s3Cache = new S3Cache(KUNTALISTA_TTL_SECONDS);
    const kuntaList: Record<string, string> = await s3Cache.get(
      "kuntalista",
      KUNTALISTA_TTL_SECONDS * 1000,
      async () => {
        log.info("Updating kuntalista, it has been expired");
        const list = await fetchKuntaLista();
        s3Cache.put("kuntalista", list);
      },
      async () => {
        log.info("Updating kuntalista, it is missing");
        const list = await fetchKuntaLista();
        s3Cache.put("kuntalista", list);
        return list;
      }
    );

    res.setHeader(
      "Cache-Control",
      "public, s-maxage=" + KUNTALISTA_TTL_SECONDS + ", stale-while-revalidate=" + (KUNTALISTA_TTL_SECONDS - 30)
    );
    res.status(200).json(kuntaList);
  });
}
