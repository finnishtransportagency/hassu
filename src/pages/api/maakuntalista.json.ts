import type { NextApiRequest, NextApiResponse } from "next";
import { S3Cache } from "../../../backend/src/cache/s3Cache";
import log from "loglevel";
import { setupLambdaMonitoring, wrapXrayAsync } from "../../../backend/src/aws/monitoring";

const MAAKUNTALISTA_TTL_SECONDS = 24 * 3 * 60 * 60;

export type MaakuntaListaOption = {
  label: string;
  labelRuo: string;
  value: string;
};

async function fetchMaakuntaLista() {
  const response = await fetch("http://rajapinnat.ymparisto.fi/api/Hakemistorajapinta/1.0/odata/Maakunta");

  const data: any = await response.json();
  const list = data?.value?.map((maakunta: any) => {
    return { label: maakunta.Nimi, labelRuo: maakunta.NimiRuo, value: maakunta.Nimi.toUpperCase() };
  });
  list.unshift({ label: "", value: "" });
  return list;
}

export default async function handler(_req: NextApiRequest, res: NextApiResponse) {
  setupLambdaMonitoring();
  return await wrapXrayAsync("handler", async () => {
    const s3Cache = new S3Cache();
    const kuntaList: Record<string, string> = await s3Cache.get(
      "maakuntalista",
      MAAKUNTALISTA_TTL_SECONDS * 1000,
      async () => {
        log.info("Updating maakuntalista, it has been expired");
        const list = await fetchMaakuntaLista();
        s3Cache.put("kuntalista", list);
      },
      async () => {
        log.info("Updating maakuntalista, it is missing");
        const list = await fetchMaakuntaLista();
        s3Cache.put("kuntalista", list);
        return list;
      }
    );

    res.setHeader(
      "Cache-Control",
      "public, s-maxage=" + MAAKUNTALISTA_TTL_SECONDS + ", stale-while-revalidate=" + (MAAKUNTALISTA_TTL_SECONDS - 30)
    );
    res.status(200).json(kuntaList);
  });
}
