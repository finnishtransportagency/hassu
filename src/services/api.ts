import getConfig from "next/config";
import { AbstractApi } from "../../common/abstractApi";

const { publicRuntimeConfig } = getConfig();
const apiImpl = publicRuntimeConfig.apiImpl;
export const api: AbstractApi = require("./api/" + apiImpl).api;

export { apiConfig } from "../../common/abstractApi";
export * from "../../common/graphql/apiModel";
