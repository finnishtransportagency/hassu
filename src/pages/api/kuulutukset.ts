import { NextApiRequest, NextApiResponse } from "next";
import { setupLambdaMonitoring, wrapXRayAsync } from "../../../backend/src/aws/monitoring";
import { Kieli } from "../../../common/graphql/apiModel";
import { ilmoitustauluSyoteHandler } from "../../../backend/src/ilmoitustauluSyote/ilmoitustauluSyoteHandler";
import { validateCredentials } from "../../util/basicAuthentication";
import { NotFoundError } from "../../../backend/src/error/NotFoundError";
import { isKieliTranslatable, KaannettavaKieli } from "common/kaannettavatKielet";

function getSingleParamValue(req: NextApiRequest, paramName: string) {
  const values = req.query[paramName];
  return values instanceof Array ? values[0] : values;
}

export default async function handler(req: NextApiRequest, res: NextApiResponse) {
  setupLambdaMonitoring();
  return await wrapXRayAsync("handler", async () => {
    if (process.env.ENVIRONMENT !== "prod" && !(await validateCredentials(req.headers.authorization))) {
      res.status(401);
      res.setHeader("www-authenticate", "Basic");

      res.send("");
      return;
    }

    const kieli = req.query["kieli"];
    if (kieli instanceof Array || !isKieliTranslatable(kieli as Kieli)) {
      res.status(400);
      res.send("");
      return;
    }

    const ely = getSingleParamValue(req, "ely");
    const lely = getSingleParamValue(req, "lely");
    const maakunta = getSingleParamValue(req, "maakunta");

    try {
      const xml = await ilmoitustauluSyoteHandler.getFeed(kieli as KaannettavaKieli, ely, lely, maakunta);
      res.setHeader("Content-Type", "application/rss+xml; charset=UTF-8");
      res.send(xml);
    } catch (e) {
      if (e instanceof NotFoundError) {
        res.status(404);
        res.send(e.message);
      } else {
        res.status(500);
        res.send(e instanceof Error ? e.message : "Virhe.");
      }
    }
  });
}
