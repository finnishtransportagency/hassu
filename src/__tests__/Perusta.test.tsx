/**
 * @jest-environment jsdom
 */

import React from "react";
import Perusta from "@pages/yllapito/perusta";
import { VelhoHakuTulos } from "@services/api";
import { create, act } from "react-test-renderer";
import { useRouter } from "next/router";

jest.mock("next/router", () => ({
  useRouter: jest.fn(),
}));

jest.mock("@services/api", () => ({
  api: {
    getVelhoSuunnitelmasByName: () => {
      const hakuTulos: VelhoHakuTulos[] = [
        { __typename: "VelhoHakuTulos", oid: "1234", nimi: "Tampereen tie Hanke", tyyppi: "Tie" },
      ];
      const promise = new Promise<VelhoHakuTulos[]>((resolve) => {
        resolve(hakuTulos);
      });
      return promise;
    },
  },
}));

describe("ProjektiTaulu", () => {
  it("renders 'Perusta' page unchanged", async () => {
    const router = {
      isReady: true,
      query: {
        projektinimi: "tampere",
      },
    };
    (useRouter as jest.Mock).mockReturnValue(router);
    const tree = create(<Perusta />);

    await act(async () => {
      await tree.update(<Perusta />);
    });
    expect(tree.toJSON()).toMatchSnapshot();
  });
});
