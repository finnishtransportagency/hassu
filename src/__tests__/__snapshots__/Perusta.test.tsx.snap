// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Perusta renders 'Perusta' page unchanged 1`] = `
Array [
  <section>
    <h1>
      Perusta uusi projekti
    </h1>
    <p
      className="ingress"
    >
      Hae projekti-VELHOon viety suunnitelma, jonka haluat tuoda Valtion väylien suunnittelu -palveluun. Voit käyttää hakuehtona projekti-VELHOon tallennettua asiatunnusta tai suunnitelman / projektin nimeä, tai näiden osaa.
    </p>
    <div
      className="w-64 content"
    >
      <form
        onSubmit={[Function]}
      >
        <div
          className="form-group"
        >
          <label>
            Asiatunnus
          </label>
          <input
            className=""
            disabled={true}
            maxLength={200}
            onChange={[Function]}
            type="text"
          />
        </div>
        <div
          className="form-group"
        >
          <label>
            Projektin nimi
          </label>
          <input
            className=""
            maxLength={100}
            name="name"
            onBlur={[Function]}
            onChange={[Function]}
            type="text"
          />
        </div>
        <div
          className="wrapper"
        >
          <button
            className="btn-primary"
            disabled={false}
          >
            <div
              className="overlay"
            >
              <div
                className="content"
              >
                Hae
                <svg
                  aria-hidden="true"
                  className="svg-inline--fa fa-search fa-w-16 ml-2 mb-px"
                  data-icon="search"
                  data-prefix="fas"
                  focusable="false"
                  role="img"
                  style={Object {}}
                  viewBox="0 0 512 512"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M505 442.7L405.3 343c-4.5-4.5-10.6-7-17-7H372c27.6-35.3 44-79.7 44-128C416 93.1 322.9 0 208 0S0 93.1 0 208s93.1 208 208 208c48.3 0 92.7-16.4 128-44v16.3c0 6.4 2.5 12.5 7 17l99.7 99.7c9.4 9.4 24.6 9.4 33.9 0l28.3-28.3c9.4-9.4 9.4-24.6.1-34zM208 336c-70.7 0-128-57.2-128-128 0-70.7 57.2-128 128-128 70.7 0 128 57.2 128 128 0 70.7-57.2 128-128 128z"
                    fill="currentColor"
                    style={Object {}}
                  />
                </svg>
              </div>
            </div>
          </button>
        </div>
      </form>
    </div>
  </section>,
  <hr />,
  <section>
    <h2>
      Hakutulokset
    </h2>
    <div
      className="info"
    >
      <div>
        <h3
          className="vayla-small-title"
        >
          Ohjeet
        </h3>
        <ul
          className="list-disc list-inside block"
        >
          <li>
            Valitse listasta se suunnitelma, jonka haluat tallentaa Valtion väylien suunnittelu -palveluun uudeksi projektiksi.
          </li>
          <li>
            Huomioi, että hakutuloksissa näytetään ainoastaan ne suunnitelmat / projektit joitaa ei ole vielä perustettu palveluun. Käytä etusivun projektihakua etsiäksesi jo perustettuja projekteja.
          </li>
        </ul>
      </div>
    </div>
    <table
      className="project-table"
    >
      <thead>
        <tr>
          <th>
            Tunnus
          </th>
          <th>
            Nimi
          </th>
          <th>
            Tyyppi
          </th>
          <th>
            Projektipäällikkö
          </th>
        </tr>
      </thead>
      <tbody>
        <tr
          className="project-row"
          href="/yllapito/projekti/1234"
          onClick={[Function]}
          onMouseEnter={[Function]}
        >
          <td
            className="sm:w-2/12"
            data-label="Tunnus"
          >
            1234
          </td>
          <td
            className="sm:w-6/12"
            data-label="Nimi"
          >
            Tampereen tie Hanke
          </td>
          <td
            className="sm:w-2/12"
            data-label="Tyyppi"
          >
            Tiesuunnitelma
          </td>
          <td
            className="sm:w-2/12"
            data-label="Projektipäällikkö"
          >
            -
          </td>
        </tr>
      </tbody>
    </table>
  </section>,
]
`;
